File: main/src/main/java/com/sedmelluq/discord/lavaplayer/tools/io/BitBufferReader.java
package com.sedmelluq.discord.lavaplayer.tools.io;

import java.io.IOException;
import java.nio.ByteBuffer;

/**
 * Helper for reading a specific number of bits at a time from a byte buffer.
 */
public class BitBufferReader extends BitStreamReader {
  private final ByteBuffer buffer;

  /**
   * @param buffer Byte buffer to read bytes from
   */
  public BitBufferReader(ByteBuffer buffer) {
    super(null);

    this.buffer = buffer;
  }

  @Override
  public long asLong(int bitsNeeded) {
    try {
      return super.asLong(bitsNeeded);
    } catch (IOException e) {
      throw new RuntimeException(e);
    }
  }

  @Override
  public int asInteger(int bitsNeeded) {
    try {
      return super.asInteger(bitsNeeded);
    } catch (IOException e) {
      throw new RuntimeException(e);
    }
  }

  @Override
  protected int readByte() throws IOException {
    return buffer.get() & 0xFF;
  }
}

File: main/src/main/java/lavaplayer/tools/io/BitBufferReader.kt
package lavaplayer.tools.io

import java.io.IOException
import java.nio.ByteBuffer
import kotlin.experimental.and

/**
 * Helper for reading a specific number of bits at a time from a byte buffer.
 * @param buffer Byte buffer to read bytes from
 */
class BitBufferReader(private val buffer: ByteBuffer) : BitStreamReader(null) {
    override fun asLong(bitsNeeded: Int): Long {
        try {
            return super.asLong(bitsNeeded)
        } catch (e: IOException) {
            throw RuntimeException(e)
        }
    }

    override fun asInteger(bitsNeeded: Int): Int {
        try {
            return super.asInteger(bitsNeeded)
        } catch (e: IOException) {
            throw RuntimeException(e)
        }
    }

    override fun readByte(): Int {
        return (buffer.get() and 0xff.toByte()).toInt()
    }
}
